planner/core/logical_plan_builder.go

```diff
@@ -5890,8 +5890,37 @@ func unfoldSelectList(list *ast.SetOprSelectList, unfoldList *ast.SetOprSelectLi
 // If asName is true, extract AsName prior to OrigName.
 // Privilege check should use OrigName, while expression may use AsName.
 // TODO: extracting all tables by vistor model maybe a better way
-func extractTableList(node ast.ResultSetNode, input []*ast.TableName, asName bool) []*ast.TableName {
+func extractTableList(node ast.Node, input []*ast.TableName, asName bool) []*ast.TableName {
 	switch x := node.(type) {
+	case *ast.DeleteStmt:
+		input = extractTableList(x.TableRefs.TableRefs, input, asName)
+		if x.IsMultiTable {
+			for _, t := range x.Tables.Tables {
+				extractTableList(t, input, asName)
+			}
+		}
+		switch w := x.Where.(type) {
+		case *ast.PatternInExpr:
+			if s, ok := w.Sel.(*ast.SubqueryExpr); ok {
+				input = extractTableList(s, input, asName)
+			}
+		case *ast.ExistsSubqueryExpr:
+			if s, ok := w.Sel.(*ast.SubqueryExpr); ok {
+				input = extractTableList(s, input, asName)
+			}
+		case *ast.BinaryOperationExpr:
+			if s, ok := w.R.(*ast.SubqueryExpr); ok {
+				input = extractTableList(s, input, asName)
+			}
+		}
```

Comment:
This part can be merged with the `select` part: `extractTableList(x.Where, input, asName)` and add 3 cases in the outer `switch-case`.